File: roisin_dubh_listings.py
#!/bin/python3
# Script to scrape the event listings webpage for the Róisín Dubh pub in Galway and generate an RSS feed
# The page does appear to offer an RSS feed but it's broken and returns dates in the far future and distant past

import requests
from datetime import datetime

listings = requests.get("https://roisindubh.net/remote/searchlistings.json").json()["results"]

# using several print statements to prioritise code readability over efficiency (I/O speeds are unimportant to me)
print('<rss xmlns:atom="http://www.w3.org/2005/Atom" version="2.0">')
print('<channel><title>Roisín Dubh Listings</title><link>https://roisindubh.net/listings/</link>')

for listing in listings:
    print('<item>')
    print('<title>' + listing["pagetitle"] + '</title>')
    print('<link>https://roisindubh.net/listings/' + listing["alias"] + '</link>')

    print('<description> <![CDATA[')
    print(listing["introtext"] + "\n" + listing["content"] + "\n")

    print('Location: ' + listing["name"] + '<br>')
    print('Ticket Allocation: ' + listing["ticket_allocation"] + '<br>')
    print('Tickets remaining?: ' + str(listing["ticket_remaining"] == '1') + '<br>')
    print('Event start time: ' + listing["event_date_time"] + '<br>')
    print('Late night?: ' + str(listing["late_night"] == "1") + '<br>')
    print('Postponed?: ' + str(listing["postponed"] == "1") + '<br>')
    print('Sales start time: ' + listing["sales_start"] + '<br>')
    print('On Sale?: ' + str(listing["on_sale"] == "1") + '<br>')
    print('Tickets remaining?: ' + str(listing["ticket_remaining"] == '1') + '<br>')

    if listing["external_ticket_url"]:
        print('External Ticket URL: <a href="' + listing["external_ticket_url"] + '">' + listing["external_ticket_url"] + '</a>')

 #  unparsed json object:
 #    "prices": {
 #      "regular": 10,
 #      "student": 0,
 #      "vip": 0,
 #      "ticket_limit": 10,
 #      "ticket_limit_per_transaction": 1,
 #      "discount": []
 #    },

    print(']]> </description>')
    print('<pubDate>' + datetime.strptime(listing["event_date_time"], '%Y-%m-%dT%H:%M:%S').strftime('%a, %d %b %Y %H:%M:%S %z') + '</pubDate>')
    print('</item>')

print('</channel></rss>')

File: roisin_dubh_listings.pl
#!/usr/bin/perl
use strict;
use warnings;

use utf8;
use JSON;
use Date;
use Date::Parse;

# use UTF-8 when writing to STDOUT
binmode(STDOUT, ":encoding(utf8)");

sub yes_or_no {
    my ($boolean) = @_;
    return($boolean eq "1" ? "Yes" : "No");
}

my $listings = decode_json(`curl "https://roisindubh.net/remote/searchlistings.json"`)->{results};

print("
<rss xmlns:atom='http://www.w3.org/2005/Atom' version='2.0'>
<channel><title>Róisín Dubh Listings</title><link>https://roisindubh.net/listings/</link>");

foreach my $listing (@$listings) {
    my $event_date = str2time($listing->{event_date_time});

    # only print data if event data is in the future
    if ($event_date > Date::now()) {
        print("
        <item>
            <title><![CDATA[" . $listing->{pagetitle} . "]]></title>
            <link>https://roisindubh.net/listings/" . $listing->{alias} . "</link>
            <pubDate>" . Date::strftime("%a, %d %b %Y %H:%M:%S %z", $event_date) . "</pubDate>

            <description>
                <![CDATA[
                    " . $listing->{introtext} . "

                    " . $listing->{content} . "

                Location: " . $listing->{name} . "<br>
                Event start time: " . Date::strftime("%Y-%m-%d %a %H:%M:%S", $event_date) . "<br>
                Late night?: " . yes_or_no($listing->{late_night}) . "<br>
                Postponed?: " . yes_or_no($listing->{postponed}) . "<br>
<br>
                Ticket Price: €" . $listing->{prices}->{regular} . "<br>
                Ticket Allocation: " . $listing->{ticket_allocation} . "<br>
                Tickets remaining?: " . yes_or_no($listing->{ticket_remaining}) . "<br>
<br>
                Sales start time: " . Date::strftime("%Y-%m-%d %a %H:%M:%S", str2time($listing->{sales_start})) . "<br>
                On Sale?: " . yes_or_no($listing->{on_sale}) . "<br>
                ]]>
            </description>
        </item>
        ");
    }
}
print("</channel></rss>");
